src/Corax/Indexing/IndexWriter.IndexEntryBuilder.cs

```diff
@@ -37,15 +39,42 @@ public void Boost(float boost)
             _parent.BoostEntry(_entryId, boost);
         }
 
-        public void Init(long entryId, int termsPerEntryIndex)
+        public void Init(long entryId, int termsPerEntryIndex, Slice documentId)
         {
             Active = true;
+            _mapFinishedSuccessfully = false;
             _entryId = entryId;
             _termPerEntryIndex = termsPerEntryIndex;
+            _documentId = documentId;
         }
 
+        /// <summary>
+        /// This method must be called before dispose to indicate there were no exceptions during indexing.
+        /// If this is not called, we will start the rollback procedure (see more at Dispose()).
+        /// This allows us to avoid using a try...catch...finally block, thereby avoiding the overhead associated with the catch block.
+        /// </summary>
+        public void EndWriting()
+        {
+            Debug.Assert(Active, "Active");
+            _mapFinishedSuccessfully = true;
+        }
+        
         public void Dispose()
         {
+            if (_mapFinishedSuccessfully == false)
```

Comment:
In what context are we _running into an error_ *and* _not rolling back the transation_ ?

Reply:
When Converter throws, e.g. for indexing complex field. If you have a map like
{"A": data, "B": data, "C": complex} without it we will index terms for "A" and "B" but never finish writing IDs (since those are indexed after all map properties} which can corrupt index.